### DISCUSSION TITLE
Running BAT script upon connecting from Moonlight

### DISCUSSION THREAD

**Discussion by judsonjames** [LizardByte/community#64] [Sunshine]
I've setup and am very satisfied with using Sunshine + Moonlight to stream games from my desktop to other places in my home. However, my main desktop is running a Samsung Odyssey G9 Neo as my only monitor, a 32:9 1440p screen. I've gone through the process and fixed my Nvidia control panel to allow my screen to run at 2560x1440p if I want.

Now, I have found success using [GameStream LaunchPad](https://github.com/cgarst/gamestream_launchpad) when running the BAT scripts needed to adjust my screen to the correct resolution/aspect ratio to comfortably stream my games at 16:9 resolutions from a 32:9 screen. However, I'm having issues getting this BAT script to run under the `Applications` section of Sunshine's UI. I have a path specified for where the script is located, and can confirm that the script starts when I supply the path to the `Command` field, but for whatever reason this starts the game display of my choice and then immediately crashes Sunshine. I don't have error codes to report since I don't know where to pull the logs. However, I then tried to follow the build-in example for Steam Big Picture Mode (BPM) and run the custom BAT as a detached command, but then it can't locate the base gamestream_launchpad.exe that's needed.

So, my question to the people of Sunshine is if anyone has a working configuration using GameStream LaunchPad or another screen size changing solution?

**Comment by ReenigneArcher** [comment]
Are you the same person that was asking about this yesterday in the discord server?

Without seeing your bat file, I am just guessing that bat is not set to use a specific working directory, and then it fails to find the executable file.

**Reply by judsonjames** [reply]
I'm not the same person as the person who was asking in the Discord.

The contents of my BAT are as follows
```bat
@ECHO OFF
start /MIN gamestream_launchpad.exe 2560 1440 gamestream_desktop.ini
```

No, I'm not specifying a specific working directory in the BAT script itself, but I assumed that it would be able to assume a working directory from the `Working Directory` parameter in the UI. Can I inject a working directory parameter in my BAT script to use with Sunshine?

**Reply by ReenigneArcher** [reply]
```batch
@ECHO OFF
cd /d <your gamestream_launchpad.exe directory>
start /MIN gamestream_launchpad.exe 2560 1440 gamestream_desktop.ini
```

Replace the the brackets with your path. If your path has spaces, you need to put double quotes (`"`) around it.

**Reply by judsonjames** [reply]
So I updated the BAT file to include the correct path as you specified. I confirmed that I can run the BAT file locally from CMD
```bat
start "" "E:\game_streaming\gamestream_launchpad\GS Playnite 1440p.bat"
```
This script works as intended. Which is fine.

However, when I put the path specified I used to run in CMD into the `Command` field of the web UI, it fails to launch the BAT with the following error in the Sunshine window

```
Warning: Couldn't run ["E:\game_streaming\gamestream_launchpad\GS Playnite 1440p.bat"]: System: The directory name is invalid
```

I also tried to use `start ""` in front of it in case it was some Windows specific error, likewise
```
Warning: Couldn't run [start "" "E:\game_streaming\gamestream_launchpad\GS Playnite 1440p.bat"]: System: The system cannot find the file specified
```

I also tried spawning the full command in the detached commands section and received this
```
Warning: Couldn't spawn [start "" "E:\game_streaming\gamestream_launchpad\GS Playnite 1440p.bat"]: System: The system cannot find the file specified
```

Am I missing something obvious?

**Reply by ReenigneArcher** [reply]
please share updated bat file

**Reply by judsonjames** [reply]
```bat
@ECHO OFF
cd /d E:\game_streaming\gamestream_launchpad
start /MIN gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini
````

**Reply by ReenigneArcher** [reply]
How about just putting this (one line) as the sunshine command?

`cd /d "E:\game_streaming\gamestream_launchpad" && start /MIN gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini`

**Reply by judsonjames** [reply]
So, apparently making it one line actually did make a difference. I updated my BAT to look like this
```bat
@ECHO OFF
cd /d "E:\game_streaming\gamestream_launchpad" && start /MIN gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini
```

And calling it from my `Detached Commands`
![image](https://user-images.githubusercontent.com/16809002/183503916-e96b63eb-fd3a-4b04-aad4-f84eaa23ed07.png)

I have also confirmed that this works with my base Desktop connection, as well as custom Steam Big Picture Mode and Launcbox/Big Box INI files

However, when I quit the target application, it does not end my connection to my desktop since it's under `Detached Commands`. I have tried to get this to run under the `Command` input on the web UI and I have not had a successful connection. The resolution is fixed from gamestream launcher but Moonlight does not end my connection there as it does with the included Steam BPM config from Sunshine.

Do I need to specify my process as the `Command` field in order to auto-disconnect from Moonshine? Or is this part of the discussion out of scope?

**Reply by ReenigneArcher** [reply]
Detached commands do not end when disconnecting.

If anything is set as a `command`, Sunshine will end once the process ends. In this case, you are starting another command window that is forgotten about immediately.

Try this:
```batch
cd /d "E:\game_streaming\gamestream_launchpad" && gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini
```

You don't need any bat file for this, seems odd to me they provide a bat file for something that should be a shortcut.

If the above works, then set your working directory to `E:\game_streaming\gamestream_launchpad` and your command to `gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini`

**Reply by judsonjames** [reply]
I think they provide the BAT because it's loadable by GeForce Experience. They don't explicitly list Sunshine as a supported option but I figured I would at least try.

However, when I updated my Sunshine UI Config to look like this
![image](https://user-images.githubusercontent.com/16809002/183510070-342242e2-06ca-4195-9de0-bcb2b656ab82.png)

It fails to connect and this message is spat out by the Sunshine console
```
Couldn't run [cd /d "E:\game_streaming\gamestream_launchpad" && gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini]: System: The system cannot find the file specified
```

**Reply by ReenigneArcher** [reply]
Try the second option I entered here: https://github.com/LizardByte/Sunshine/discussions/310#discussioncomment-3352647

**Reply by judsonjames** [reply]
I tried the second option you specified , this is the error message
```
[gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini]: System: The system cannot find the file specified
```

**Reply by KaosC57** [reply]
> I think they provide the BAT because it's loadable by GeForce Experience. They don't explicitly list Sunshine as a supported option but I figured I would at least try.
> 
> However, when I updated my Sunshine UI Config to look like this ![image](https://user-images.githubusercontent.com/16809002/183510070-342242e2-06ca-4195-9de0-bcb2b656ab82.png)
> 
> It fails to connect and this message is spat out by the Sunshine console
> 
> ```
> Couldn't run [cd /d "E:\game_streaming\gamestream_launchpad" && gamestream_launchpad.exe 2560 1440 gamestream_playnite.ini]: System: The system cannot find the file specified
> ```
![2022-08-10 21_29_49-Sunshine — Mozilla Firefox](https://user-images.githubusercontent.com/8933601/184055286-a681f5e0-cffc-4712-8c71-9e11a6aa78ab.png)

This is all I have to do to get my gamestream_launchpad to function properly in Sunshine.

Additionally, here's my Bat file contents.

`@ECHO OFF`

`cd /d "S:\Games\gamestream_launchpad"`

`start /MIN gamestream_launchpad.exe 1920 1080 gamestream_playnite.ini`

**Reply by ReenigneArcher** [reply]
@KaosC57 to avoid a bat file you could just do this `cd /d "S:\Games\gamestream_launchpad" && start /MIN gamestream_launchpad.exe 1920 1080 gamestream_playnite.ini`

Any time you use `start` you will need to use a detached command in sunshine, since the `start` process does not continue as a sub process but instead starts a new process.

**Comment by adam-fox** [comment]
@ReenigneArcher i got this working using a batch file but wasn't able to do so with the script you specified. regardless, working fine with the batch file. question, though: how might i go about having sunshine detect when gamestream_launchpad.exe exits and end the stream?

[a comment by an automated account is not shown]
